container/openejb-core/src/main/java/org/apache/openejb/util/PropertyPlaceHolderHelper.java

```diff
@@ -53,33 +55,34 @@ public static void reset() {
     }
 
     public static String simpleValue(final String raw) {
-        if (raw == null) {
-            return null;
-        }
-        if (!raw.contains(PREFIX) || !raw.contains(SUFFIX)) {
-            return String.class.cast(decryptIfNeeded(raw.replace(PREFIX, "").replace(SUFFIX, ""), false));
-        }
-
-        String value = SUBSTITUTOR.replace(raw);
-        if (!value.equals(raw) && value.startsWith("java:")) {
-            value = value.substring(5);
-        }
-        return String.class.cast(decryptIfNeeded(value.replace(PREFIX, "").replace(SUFFIX, ""), false));
+        return String.class.cast(simpleValueAsStringOrCharArray(raw, false));
     }
 
     public static Object simpleValueAsStringOrCharArray(final String raw) {
+        return simpleValueAsStringOrCharArray(raw, true);
+    }
+
+    private static Object simpleValueAsStringOrCharArray(final String raw, boolean acceptCharArray) {
         if (raw == null) {
             return null;
         }
         if (!raw.contains(PREFIX) || !raw.contains(SUFFIX)) {
-            return decryptIfNeeded(raw.replace(PREFIX, "").replace(SUFFIX, ""), true);
+            return decryptIfNeeded(raw, acceptCharArray);
         }
 
         String value = SUBSTITUTOR.replace(raw);
+
         if (!value.equals(raw) && value.startsWith("java:")) {
             value = value.substring(5);
         }
-        return decryptIfNeeded(value.replace(PREFIX, "").replace(SUFFIX, ""), true);
+
+        // no key defined and no escape sequence found -> substitute to key instead
+        if (!(raw.contains(ESCAPE_SEQUENCE) || value.contains(ESCAPE_SEQUENCE))
```

Comment:
I'm not sure you can implement this with "contains" checks (only t"here is none of these chars" test can be reliable) since you need the presence or not of the char + the position between the present ones to have a reliable test otherwise you can fall back in the same issue you originall hit which is to call "replace" where it is not needed.

Test are missing all these weirdo cases but since it is more or less the one you hit I guess it can happen.

A probably trivial fix is to extend PropertiesLookup to handle the fallback replacement (instead of null it returns the key) and just always call the substitutor if there is at least start+end chars. Worse case the substitutor will already do nothing.

Wdyt?

Reply:
Thx @rmannibucau  for the follow up!

Yes - testing with `contains` is a simplification and neglects some weirdo cases... and yes: it would require char, position + offsets to properly test and replace the closing and opening brackets. 

Thus, I like the idea of using `PropertiesLookup#lookup` and returning the key there. The parsing "magic" (char, position, offsets) is already done in the substitutor, so I guess fallbacking there is a good option. 

I will take a look at it.

Reply:
@rmannibucau 

Wouldn't returning the `key` instead of `null` in `PropertiesLookup#lookup` breaking the `:-` syntax? 

From my understanding, `:-` is used for some sort of inlined "if else", i.e. "if foo is not defined, use bar instead" - is this assumption correct? Or maybe I didn't understood the substition process for `:-` correctly? :)

For example:

- "${foo:-bar}" -> `bar` (in the original implementation (returning `null`); `foo` is undefined)
- "${foo:-bar}" -> `foo` (if we return the `key` instead of `null`, the second part via  `:-` is not evaluated as we have a match for the given key (which is the key itself))